File: polyglot/overwrite_from_context/mjs/src/compute-registry.ts
import { Logger } from './logger.js';
import { ComputeScope } from './options.js';
import { ComputeFunctionError, ErrorCode } from './errors.js';

type ComputeFunction = (context?: any) => any | Promise<any>;

interface RegisteredFunction {
    fn: ComputeFunction;
    scope: ComputeScope;
}

export class ComputeRegistry {
    private logger: Logger;
    private functions: Map<string, RegisteredFunction>;
    private cache: Map<string, any>;
    private static NAME_PATTERN = /^[a-zA-Z_][a-zA-Z0-9_]*$/;

    constructor(logger?: Logger) {
        this.logger = logger || Logger.create('runtime-template-resolver', 'compute-registry.ts'); // Using static filename as __filename is not available in ESM easily without helpers
        this.functions = new Map();
        this.cache = new Map();
        this.logger.debug('ComputeRegistry initialized');
    }

    public register(name: string, fn: ComputeFunction, scope: ComputeScope): void {
        this.validateName(name);
        this.logger.debug(`Registering function: ${name} with scope: ${scope}`);
        this.functions.set(name, { fn, scope });
        this.logger.info(`Function registered: ${name}`);
    }

    public unregister(name: string): void {
        if (this.functions.has(name)) {
            this.logger.debug(`Unregistering function: ${name}`);
            this.functions.delete(name);
            this.logger.info(`Function unregistered: ${name}`);
        }
    }

    public has(name: string): boolean {
        return this.functions.has(name);
    }

    public list(): string[] {
        return Array.from(this.functions.keys());
    }

    public getScope(name: string): ComputeScope | undefined {
        return this.functions.get(name)?.scope;
    }

    public clear(): void {
        this.logger.debug('Clearing registry');
        this.functions.clear();
        this.cache.clear();
    }

    public clearCache(): void {
        this.logger.debug('Clearing result cache');
        this.cache.clear();
    }

    public async resolve(name: string, context?: any): Promise<any> {
        this.logger.debug(`Resolving function: ${name}`);

        const regFn = this.functions.get(name);
        if (!regFn) {
            throw new ComputeFunctionError(
                `Compute function not found: ${name}`,
                ErrorCode.COMPUTE_FUNCTION_NOT_FOUND,
                { name }
            );
        }

        // Check cache for STARTUP functions
        if (regFn.scope === ComputeScope.STARTUP && this.cache.has(name)) {
            this.logger.debug(`Returning cached value for: ${name}`);
            return this.cache.get(name);
        }

        try {
            const result = await regFn.fn(context);

            // Cache result if STARTUP scope
            if (regFn.scope === ComputeScope.STARTUP) {
                this.cache.set(name, result);
            }

            return result;
        } catch (e: any) {
            this.logger.error(`Function execution failed: ${name}, error: ${e.message}`);
            throw new ComputeFunctionError(
                `Compute function failed: ${name}`,
                ErrorCode.COMPUTE_FUNCTION_FAILED,
                { name, originalError: e.message }
            );
        }
    }

    private validateName(name: string): void {
        if (!name) {
            throw new Error('Function name cannot be empty');
        }
        if (!ComputeRegistry.NAME_PATTERN.test(name)) {
            throw new Error(`Invalid function name: ${name}. Must match pattern: ^[a-zA-Z_][a-zA-Z0-9_]*$`);
        }
    }
}

File: polyglot/overwrite_from_context/py/src/runtime_template_resolver/compute_registry.py
import re
import asyncio
from typing import Callable, Dict, Optional, Any, List, Union
from dataclasses import dataclass

from .logger import Logger
from .options import ComputeScope
from .errors import ComputeFunctionError, ErrorCode

@dataclass
class RegisteredFunction:
    fn: Callable
    scope: ComputeScope

class ComputeRegistry:
    NAME_PATTERN = re.compile(r'^[a-zA-Z_][a-zA-Z0-9_]*$')

    def __init__(self, logger: Optional[Logger] = None):
        self._logger = logger or Logger.create("runtime_template_resolver", __file__)
        self._functions: Dict[str, RegisteredFunction] = {}
        self._cache: Dict[str, Any] = {}
        self._logger.debug("ComputeRegistry initialized")

    def register(self, name: str, fn: Callable, scope: ComputeScope) -> None:
        self._validate_name(name)
        self._logger.debug(f"Registering function: {name} with scope: {scope}")
        self._functions[name] = RegisteredFunction(fn=fn, scope=scope)
        self._logger.info(f"Function registered: {name}")

    def unregister(self, name: str) -> None:
        if name in self._functions:
            self._logger.debug(f"Unregistering function: {name}")
            del self._functions[name]
            self._logger.info(f"Function unregistered: {name}")

    def has(self, name: str) -> bool:
        return name in self._functions

    def list(self) -> List[str]:
        return list(self._functions.keys())

    def get_scope(self, name: str) -> Optional[ComputeScope]:
        if name in self._functions:
            return self._functions[name].scope
        return None

    def clear(self) -> None:
        self._logger.debug("Clearing registry")
        self._functions.clear()
        self._cache.clear()

    def clear_cache(self) -> None:
        self._logger.debug("Clearing result cache")
        self._cache.clear()

    async def resolve(self, name: str, context: Optional[Dict[str, Any]] = None) -> Any:
        self._logger.debug(f"Resolving function: {name}")
        
        if name not in self._functions:
            raise ComputeFunctionError(
                f"Compute function not found: {name}",
                ErrorCode.COMPUTE_FUNCTION_NOT_FOUND,
                {"name": name}
            )

        reg_fn = self._functions[name]

        # Check cache for STARTUP functions
        if reg_fn.scope == ComputeScope.STARTUP and name in self._cache:
            self._logger.debug(f"Returning cached value for: {name}")
            return self._cache[name]

        try:
            # Check if function expects arguments
            # Simple check for now, can be more robust with signature inspection
            # Assuming fn(context) or fn()
            
            result = None
            if asyncio.iscoroutinefunction(reg_fn.fn):
                try:
                    result = await reg_fn.fn(context)
                except TypeError:
                     # Fallback if it doesn't accept context
                    result = await reg_fn.fn()
            else:
                try:
                    result = reg_fn.fn(context)
                except TypeError:
                     # Fallback if it doesn't accept context
                    result = reg_fn.fn()
            
            # Cache result if STARTUP scope
            if reg_fn.scope == ComputeScope.STARTUP:
                self._cache[name] = result

            return result

        except Exception as e:
            self._logger.error(f"Function execution failed: {name}, error: {str(e)}")
            raise ComputeFunctionError(
                f"Compute function failed: {name}",
                ErrorCode.COMPUTE_FUNCTION_FAILED,
                {"name": name, "original_error": str(e)}
            ) from e

    def _validate_name(self, name: str) -> None:
        if not name:
             raise ValueError("Function name cannot be empty")
        if not self.NAME_PATTERN.match(name):
            raise ValueError(f"Invalid function name: {name}. Must match pattern: ^[a-zA-Z_][a-zA-Z0-9_]*$")
